File: src/main/java/me/meiallu/luck/run/Parser.java
package me.meiallu.luck.run;

import me.meiallu.luck.data.Token;
import me.meiallu.luck.data.TokenType;
import me.meiallu.luck.function.Functions;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;

public class Parser {

    private static HashMap<String, Object> variables = new HashMap<>();

    public static void run(List<Token> tokenList) {
        for (int i = 0; i < tokenList.size(); ) {
            Token token = tokenList.get(i);

            switch (token.type) {
                case TYPE:
                    String name = tokenList.get(++i).value.toString();
                    Object value = null;

                    if (tokenList.get(i + 1).type == TokenType.ASSIGNMENT_OPERATOR)
                        value = evaluate(tokenList, i += 2);

                    variables.put(name, value);
                    break;
                case FUNCTION:
                    List<List<Token>> arguments = new ArrayList<>();
                    Token loopedToken = tokenList.get(++i);

                    while (loopedToken.type != TokenType.FUNCTION_END && loopedToken.type != TokenType.NEW_LINE) {
                        List<Token> currentArgument = new ArrayList<>();

                        while (loopedToken.type != TokenType.SEPARATOR && loopedToken.type != TokenType.FUNCTION_END) {
                            currentArgument.add(loopedToken);
                            loopedToken = tokenList.get(++i);
                        }

                        if (i < tokenList.size() - 2)
                            loopedToken = tokenList.get(++i);
                        if (!currentArgument.isEmpty())
                            arguments.add(currentArgument);
                    }

                    Object[] objects = new Object[arguments.size()];
                    int index = 0;

                    for (List<Token> tokens : arguments) {
                        objects[index] = evaluate(tokens, 0);
                        index++;
                    }

                    Functions.get((String) token.value).call(objects);
                    break;
            }

            i++;
        }
    }

    private static Object evaluate(List<Token> tokenList, int currentIndex) {
        Object value = null;

        TokenType type = tokenList.get(currentIndex).type;
        List<Token> assignment;

        if (currentIndex == 0) {
            assignment = tokenList;
        } else {
            int lastIndex = currentIndex;

            while (lastIndex < tokenList.size() - 1 && type != TokenType.NEW_LINE && type != TokenType.SEPARATOR && type != TokenType.FUNCTION_END)
                type = tokenList.get(++lastIndex).type;

            assignment = tokenList.subList(currentIndex, lastIndex);
        }

        if (isExpression(assignment)) {
            Iterator<Token> iterator = assignment.iterator();

            while (iterator.hasNext()) {
                Token token = iterator.next();

                if (value == null) {
                    value = switch (token.type) {
                        case NUMBER -> (Double) token.value;
                        case IDENTIFIER -> (Double) variables.get((String) token.value);
                        default -> throw new IllegalStateException("Unexpected value: " + token.type);
                    };
                    continue;
                }

                if (token.type == TokenType.ARITHMETIC_OPERATOR) {
                    Token nextToken = iterator.next();

                    Double nextDouble = switch (nextToken.type) {
                        case NUMBER -> (Double) nextToken.value;
                        case IDENTIFIER -> (Double) variables.get((String) nextToken.value);
                        default -> throw new IllegalStateException("Unexpected type: " + nextToken.type);
                    };

                    value = switch ((char) token.value) {
                        case '+' -> (Double) value + nextDouble;
                        case '-' -> (Double) value - nextDouble;
                        case '*' -> (Double) value * nextDouble;
                        case '/' -> (Double) value / nextDouble;
                        case '%' -> (Double) value % nextDouble;
                        default -> throw new IllegalStateException("Unexpected operator: " + (char) token.value);
                    };
                }
            }
        } else if (assignment.get(0).type == TokenType.IDENTIFIER) {
            value = variables.get(assignment.get(0).value.toString());
        } else {
            value = assignment.get(0).value;
        }

        return value;
    }

    private static boolean isExpression(List<Token> tokenList) {
        for (Token token : tokenList)
            if (token.type == TokenType.ARITHMETIC_OPERATOR)
                return true;

        return false;
    }
}

File: src/main/kotlin/Parser.kt
package me.meiallu.luck

class Parser {

    companion object {
        fun run(tokenList: ArrayList<Token>) {
            var i = 0

            while (i < tokenList.size) {
                val token = tokenList[i]

                when (token.type) {
                    TokenType.TYPE -> {
                        val name = tokenList[++i].value as String
                        var value: Any? = null

                        if (tokenList[i + 1].type == TokenType.ASSIGNMENT_OPERATOR) {
                            i += 2
                            value = evaluate(tokenList, i)
                        }

                        variables[name] = value as Any
                    }

                    TokenType.IDENTIFIER -> {
                        val identifier = token.value as String

                        if (variables.containsKey(identifier) && tokenList[i + 1].type == TokenType.ASSIGNMENT_OPERATOR) {
                            i += 2
                            val identifierValue = evaluate(tokenList, i)

                            variables[identifier] = identifierValue
                        }
                    }

                    TokenType.FUNCTION -> runFunction(tokenList, token.value as String, i)
                    else -> {}
                }

                i++
            }
        }

        private fun runFunction(tokenList: MutableList<Token>, functionName: String, currentIndex: Int): Any? {
            var index: Int = currentIndex

            val arguments: ArrayList<ArrayList<Token>> = ArrayList()
            var loopedToken = tokenList[++index]

            while (loopedToken.type != TokenType.FUNCTION_END && loopedToken.type != TokenType.NEW_LINE) {
                val currentArgument: ArrayList<Token> = ArrayList()

                while (loopedToken.type != TokenType.SEPARATOR && loopedToken.type != TokenType.FUNCTION_END) {
                    currentArgument += loopedToken
                    loopedToken = tokenList[++index]
                }

                if (index < tokenList.size - 2)
                    loopedToken = tokenList[++index]
                if (currentArgument.isNotEmpty())
                    arguments += currentArgument
            }

            val objects = arrayOfNulls<Any>(arguments.size)

            for (i in 0..<arguments.size)
                objects[i] = evaluate(arguments[i], 0)

            return Functions.get(functionName)?.call(objects)
        }

        private fun evaluate(tokenList: ArrayList<Token>, currentIndex: Int): Any {
            var assignmentList: MutableList<Token> = tokenList

            if (currentIndex != 0) {
                var type = tokenList[currentIndex].type
                var lastIndex = currentIndex

                while (type != TokenType.NEW_LINE && type != TokenType.SEPARATOR && type != TokenType.FUNCTION_END)
                    type = tokenList[++lastIndex].type

                assignmentList = tokenList.subList(currentIndex, lastIndex)
            }

            if (isExpression(assignmentList)) {
                val iterator = assignmentList.iterator()
                var token = iterator.next()

                var value = when (token.type) {
                    TokenType.NUMBER -> token.value as Double
                    TokenType.IDENTIFIER -> variables[token.value as String] as Double
                    else -> throw IllegalArgumentException("Unexpected type: " + token.type)
                }

                while (iterator.hasNext()) {
                    token = iterator.next()

                    if (token.type == TokenType.ARITHMETIC_OPERATOR) {
                        val nextToken = iterator.next()

                        val nextDouble = when (nextToken.type) {
                            TokenType.NUMBER -> nextToken.value as Double
                            TokenType.IDENTIFIER -> variables[nextToken.value as String] as Double
                            else -> throw IllegalStateException("Unexpected type: " + nextToken.type)
                        }

                        value = when (token.value as Char) {
                            '+' -> value + nextDouble
                            '-' -> value - nextDouble
                            '*' -> value * nextDouble
                            '/' -> value / nextDouble
                            '%' -> value % nextDouble
                            else -> throw IllegalStateException("Unexpected operator: " + token.value as Char)
                        }
                    }
                }

                return value
            }

            return when (assignmentList[0].type) {
                TokenType.IDENTIFIER -> variables[assignmentList[0].value as String]
                TokenType.FUNCTION -> {
                    assignmentList += Token(TokenType.FUNCTION_END, null)
                    runFunction(assignmentList, assignmentList[0].value as String, 0)
                }

                TokenType.STRING, TokenType.NUMBER -> assignmentList[0].value
                else -> null
            } ?: "null"
        }

        private fun isExpression(tokenList: MutableList<Token>): Boolean {
            for (token in tokenList)
                if (token.type == TokenType.ARITHMETIC_OPERATOR)
                    return true

            return false
        }
    }
}